application/controllers/admin/Database.php

```diff
@@ -277,571 +224,95 @@ private function actionUpdateDefaultValues($iSurveyID)
-//                                    $sError = '<strong>'.sprintf(gT('A code has been updated to %s.'), $oSubQuestion->title).'</strong><br/>';
-//                                    Yii::app()->setFlashMessage($sError, 'error');
-//                                    $bAnswerSave = true;
-//                                }
-//                            }
-//                        } else {
-//                            Yii::app()->setFlashMessage(sprintf(gT("Subquestions %s could not be updated."), $aCodes[$iScaleID][$iPosition]), 'error');
-//                        }
-//                    }
-//                    $iPosition++;
-//                }
-//            }
-//        }
-//
-//        //This is SUPER important! Recalculating the ExpressionScript Engine state!
-//        LimeExpressionManager::SetDirtyFlag();
-//        LimeExpressionManager::UpgradeConditionsToRelevance($iSurveyID);
-//        $this->resetEM();
-//
-//        if (!isset($aErrors) || !count($aErrors)) {
-//            if (!Yii::app()->request->getPost('bFullPOST')) {
-//                Yii::app()->session['flashmessage'] = gT("Not all subquestions were saved. This usually happens due to server limitations ( PHP setting max_input_vars) - please contact your system administrator.");
-//            } else {
-//                Yii::app()->session['flashmessage'] = gT("Subquestions were successfully saved.");
-//            }
-//        }
-//
-//        if (Yii::app()->request->getPost('close-after-save') === 'true') {
-//            $this->getController()->redirect(array('questionEditor/view/surveyid/'.$iSurveyID.'/gid/'.$this->iQuestionGroupID.'/qid/'.$this->iQuestionID));
-//        }
-//
-//        $this->getController()->redirect(array('/admin/questions/sa/subquestions/surveyid/'.$iSurveyID.'/gid/'.$this->iQuestionGroupID.'/qid/'.$this->iQuestionID));
-//    }
-
-    /**
-     * action to do when update survey seetings + survey language
+     * Action to run when update survey settings + survey language
+     *
+     * Refactored to use Services\SurveyUpdater 2023-05-30 (kfoster).
+     *
      * @param integer $iSurveyID
+     * @param ?array $input For dependency injection during testing
      * @return void (redirect)
      */
-    private function actionUpdateSurveyLocaleSettings($iSurveyID)
+    private function actionUpdateSurveyLocaleSettings($surveyId, $input = [])
     {
+        $diContainer = \LimeSurvey\DI::getContainer();
+        $surveyUpdater = $diContainer->get(
+            LimeSurvey\Models\Services\SurveyUpdater::class
+        );
 
-        //@todo  here is something wrong ...
-        $oSurvey = Survey::model()->findByPk($iSurveyID);
-        $languagelist = $oSurvey->additionalLanguages;
-        $languagelist[] = $oSurvey->language;
-
-        Yii::app()->loadHelper('database');
-
-        $hasSurveyLanguageSettingError = false;
-        if (Permission::model()->hasSurveyPermission($iSurveyID, 'surveylocale', 'update')) {
```

Comment:
Permission checks should still be in the controller, IMO, and not service classes.

Reply:
Logic around who can and can not perform certain actions is business logic.

Controllers need to handle permission failure in its own specific way which is why the service throws a permission failure exception.

Reply:
Lemme think tho

Reply:
What if you wanna use that service class from a CLI command, where no permissions should be checked at all?

Reply:
Well you could inject a dummy permission checker, I guess.